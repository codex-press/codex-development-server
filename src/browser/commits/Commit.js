import React from 'react';

import './Commit.less';

export default function Commit(props) {

      /*
      <h4>
        { props.repository_name } &ndash; {}
        <span className="sha">{ props.sha.substr(0, 5) }</span>
      </h4>
      */

  console.log(props);

  return (
    <div className="Commit">

      <div className="message">
        { props.message }
      </div>

      <div className="progress-bar">
        <div style={ { width: (props.build_progress * 100 + '%') } }>
          { Math.round(props.build_progress * 100) + '%' }
        </div>
      </div> 

      <pre>
        { props.build_output.join('') }
      </pre>

    </div>
  );

}


