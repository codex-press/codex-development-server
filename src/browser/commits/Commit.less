@import (reference) "../index.less";

.Commit {
  flex: 1 1 400px;
  min-height: 150px;
  display: flex;
  flex-flow: column nowrap;


  .BuildOutput.minimal & {
    flex: 1 1 auto;
    min-height: auto;

    .message { 
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    pre {
      display: none;
    }

    .message {
      display: flex;
      flex-flow: row nowrap;
    }

    .message *:not(:last-child) {
      margin-left: 0.2em;
    }

    .message .commit-link {
      flex: 1 1 auto;
      min-width: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

  }

  .repository-name {
    font-weight: 800;
  }

  a {
    color: inherit;
    text-decoration: none;
  }

  a:hover {
    text-decoration: underline;
  }

  .sha {
    font-family: @mono;    
  }

  .progress-bar {
    margin: 0.5rem 0;
  }

  pre {
    flex: 1 1 auto;
    font-family: @mono;
    border-radius: 4px;
    margin: 1rem 0;
    background: #ddd;
    color: #333;
    padding: 0.5rem;
    overflow: scroll;
  }

}

.progress-bar {
  height: 1.3rem;
  background-color: #ddd;
  border-radius: 0.2rem;


  .bar {
    min-width: 2.2em;
    color: white;
    font-weight: 500;
    text-align: right;
    padding: 0 0.3em;
    transition: width 0.3s ease;
    height: 100%;
    border-radius: inherit;
    position: relative;
    overflow: hidden;
  }

  .text {
    position: relative;
    z-index: 3;  
  }

  .stripes {
    position: absolute;
    left: -40px;
    width: calc(100% + 50px);
    height: 100%;
    background-color: hsl(206, 73%, 47%);
    background-image: linear-gradient(
      45deg,
      rgba(255,255,255,.15) 25%,
      transparent 25%,
      transparent 50%,
      rgba(255,255,255,.15) 50%,
      rgba(255,255,255,.15) 75%,
      transparent 75%,
      transparent
    );
    background-size: 40px 40px;
    animation: progress-bar-stripes 2s linear infinite;
  }


  @keyframes progress-bar-stripes {
    from { transform: translate(0); }
    to   { transform: translate(40px); }
  }


  .indicator {
    position: absolute;
    content: '';
    opacity: 0;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    transform: scale(0, 1);
    transform-origin: 0;
    background: #fff;
    border-radius: inherit;
    animation: progress-active 2s ease infinite;
  }


  @keyframes progress-active {
    from {
      opacity: 0.3;
      transform: scale(0, 1);
    }
    to {
      opacity: 0;
      transform: scale(1);
    }
  }

}

