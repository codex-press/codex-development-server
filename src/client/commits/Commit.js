import React from 'react';
import * as env from '../env';

import './Commit.less';

export default function Commit(props) {

  // This is likely because we got a message but have not fetched corresponding
  // data from the server yet
  if (!props.repository_name)
    return null;

  const repoLink = env.codexOrigin + '/edit/repos/' + props.repository_name
  const commitLink = repoLink + '/' + props.sha

  const wheel = e => {
    e.preventDefault()
    e.target.scrollTop += e.deltaY
  }

  const scrollBottom = el => el && (el.scrollTop = el.scrollHeight)

  const pct = Math.round(props.build_progress * 100) + '%'

  return (
    <div className="Commit">

      <div className="message">

        <a className="repository-name" href={ repoLink } target="_blank">
          { props.repository_name }:
        </a>
        {' '}
        <a className="commit-link" href={ commitLink } target="_blank">
          { props.message.trim() }
        </a>
        {' '}
        { props.build_progress == 1 &&
          <button className="small" onClick={ props.clear }>
            Clear
          </button>
        }
      </div>

      { props.build_progress < 1 &&
        <div className="progress-bar">
          <div className="bar" style={ { width: pct } }>
            <div className="stripes" />
            <div className="indicator" />
            <div className="text">{ pct }</div>
          </div>
        </div> 
      }

      <pre onWheel={ wheel } ref={ scrollBottom }>
        { props.build_output instanceof Array ? 
          props.build_output.join('') :
          props.build_output
        }
      </pre>

    </div>
  );

}


