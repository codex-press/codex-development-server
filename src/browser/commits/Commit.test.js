import React from 'react'
import { shallow } from 'enzyme'

import Commit from './Commit'

const commit = {
  permission: true,
  repository_name: 'lib',
  id: "004b6ed6-4106-42be-a975-89f7cb2ddfa6",
  repository_id: 'f2e8b08a-aceb-415a-9ff8-3efe1bc5757f',
  sha: '16ad6335d2facd9a0f3e600cca9c77bf61017cb5',
  branch: 'master',
  build_status: 'pending',
  build_output: null,
  message: 'changing this all around. open_iconic now just SVG\n',
  author_name: 'Tim Farnam',
  build_progress: '0.55',
  build_output: [ 'first line\n', 'second line\n' ],
  author_email: 'twfarnam@gmail.com',
  created_at: "2017-07-01T22:47:58.810Z",
}


test('renders Commit', () => {
  const c = shallow(<Commit { ...commit } />)
  expect(c).toMatchSnapshot()
})


