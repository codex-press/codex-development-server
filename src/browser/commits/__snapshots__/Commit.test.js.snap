// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Commit 1`] = `
<div
  className="Commit"
>
  <div
    className="message"
  >
    <a
      className="repository-name"
      href="https://codex.press/edit/repos/lib"
      target="_blank"
    >
      lib
      :
    </a>
     
    <a
      className="commit-link"
      href="https://codex.press/edit/repos/lib/16ad6335d2facd9a0f3e600cca9c77bf61017cb5"
      target="_blank"
    >
      changing this all around. open_iconic now just SVG
    </a>
     
  </div>
  <div
    className="progress-bar"
  >
    <div
      className="bar"
      style={
        Object {
          "width": "55%",
        }
      }
    >
      <div
        className="stripes"
      />
      <div
        className="indicator"
      />
      <div
        className="text"
      >
        55%
      </div>
    </div>
  </div>
  <pre
    onWheel={[Function]}
  >
    first line
    second line
    
  </pre>
</div>
`;
